plugins/cn0540.c

```diff
@@ -0,0 +1,575 @@
+	set_voltage(dac_ch, (double)(vshift_mv / DAC_BUF_GAIN));
+
+	g_free(vshift_string);
+	fflush(stdout);
+}
+
+static void read_vsensor(GtkButton *btn)
+{
+
+	char buff_string[9];
+	double vsensor_mv, vshift_mv,vadc_mv;
+
+	vadc_mv = get_voltage(adc_ch);
+	vshift_mv = get_vshift_mv();
+	double v1_st = FDA_VOCM_MV - vadc_mv/FDA_GAIN;
+	vsensor_mv = (((G + 1) *vshift_mv ) - v1_st )/G;
+	vsensor_mv -= get_voltage(adc_ch);
+
+	snprintf(buff_string, sizeof(buff_string), "%f", vsensor_mv);
+	gtk_text_buffer_set_text(vsensor_buf, buff_string, -1);
+}
+
+static void calib(GtkButton *btn)
+{
+	double adc_voltage_mv, dac_voltage_mv;
+	char buff_string[9];
+	int i;
+
+	gtk_text_buffer_set_text(calib_buffer, "Calibrating...", -1);
+
+	for(i = 0; i < CALIB_MAX_ITER; i ++)
+	{
+		adc_voltage_mv = get_voltage(adc_ch);
+		dac_voltage_mv = get_voltage(dac_ch) - adc_voltage_mv;
+		set_voltage(dac_ch, dac_voltage_mv);
+		delay_ms(10);
+	}
+	snprintf(buff_string, sizeof(buff_string), "%f", adc_voltage_mv);
+	gtk_text_buffer_set_text(calib_buffer, buff_string, -1);
+	gtk_button_clicked(read_btn);
+	gtk_button_clicked(readvsensor_btn);
+}
+
+ static void save_widget_value(GtkWidget *widget, struct iio_widget *iio_w)
+ {
+	iio_w->save(iio_w);
+	/* refresh widgets so that, we know if our value was updated */
+	iio_update_widgets(iio_widgets, num_widgets);
+ }
+
+ static void make_widget_update_signal_based(struct iio_widget *widgets,
+						unsigned int num_widgets)
+ {
+	char signal_name[25];
+	unsigned int i;
+
+	for (i = 0; i < num_widgets; i++) {
+		if (GTK_IS_SPIN_BUTTON(widgets[i].widget) &&
+				widgets[i].priv_progress != NULL) {
+			iio_spin_button_progress_activate(&widgets[i]);
+		} else {
+			g_signal_connect(G_OBJECT(widgets[i].widget),
+					 signal_name,
+					 G_CALLBACK(save_widget_value),
+					 &widgets[i]);
+		}
+	}
+ }
+
+static void cn0540_get_channels()
+{
+	gboolean direction = TRUE;
+	int idx = -1;
+	char *label;
+
+	adc_ch = iio_device_find_channel(iio_adc, ADC_DEVICE_CH, FALSE);
+	dac_ch = iio_device_find_channel(iio_dac, DAC_DEVICE_CH, TRUE);
+
+	label = strdup(GPIO_CTRL_FIRST_CH);
+	while(1) {
+		gpio_ch[++idx].gpio = iio_device_find_channel(iio_gpio, label,
+								  direction);
+		if (gpio_ch[idx].gpio != NULL){
+			iio_channel_attr_read(gpio_ch[idx].gpio, "label",
+						  gpio_ch[idx].label, 30);
+			label[7]++;
+		} else if (direction && (gpio_ch[idx].gpio == NULL)) {
+			direction = !direction;
+			label = strdup(GPIO_CTRL_FIRST_CH);
+			idx--;
+		} else {
+			break;
+		}
+	}
+
+	if(iio_voltage_mon != NULL) {
+		if(!strcmp(iio_device_get_name(iio_voltage_mon),
+			   VOLTAGE_MONITOR_1)) {
+			label = strdup(VOLTAGE_MONITOR_1_FIRST_CH);
+		} else {
+			label = strdup(VOLTAGE_MONITOR_2_FIRST_CH);
+		}
+		for(idx = 0; idx < NUM_ANALOG_PINS; idx++) {
+			analog_in[idx] = iio_device_find_channel(
+						iio_voltage_mon, label, FALSE);
+			label[strlen(label) - 1]++;
+			if (label[7] == ':')
+				label = strdup("voltage10");
+		}
+	}
+}
+
```

Comment:
hmmm, `strdup()` returns a a new string which is obtained with `malloc()`. I can see a lot of leaks in this function...

Reply:
@scuciurean please fix this so we can merge this